File: MaxSumCircularSubarray.java
class MaxSumCircularSubarray {
    public int maxSubarraySumCircular(int[] nums) {
        int total = 0, curMax = 0, curMin = 0;
        int gloMax = nums[0], gloMin = nums[0];
        
        for (int n : nums) {
            curMax = Math.max(curMax + n, n); 
            gloMax = Math.max(gloMax, curMax); 
            
            curMin = Math.min(curMin + n, n);
            gloMin = Math.min(gloMin, curMin); 
            
            total += n; 
        }
        
        
        if (gloMax > 0) {
            return Math.max(gloMax, total - gloMin);
        } else {
            return gloMax;
        }
    }
}

File: MaxSumCircularSubarray.py
# https://leetcode.com/problems/maximum-sum-circular-subarray/description/
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        total=0
        curMax,curMin=0,0
        gloMax,gloMin=nums[0],nums[0]

        for n in nums:
            curMax=max(curMax+n,n)
            curMin=min(curMin+n,n)
            total+=n
            gloMax=max(gloMax,curMax)
            gloMin=min(gloMin,curMin)
        return max(gloMax,total-gloMin) if gloMax>0 else gloMax

